### PR TITLE
use ICU

### PR DESCRIPTION
**Pull request by mr-martian**
ICU changes
- replace all uses of `std::wstring` and associated types with `UString`
- use `lttoolbox/input_file.h` to read UTF-8 stream with nulls, since ICU can't do that directly
- use lttoolbox helper functions for manipulating string case

efficiency/readability changes
- pass containers by `const` reference whenever possible
- in `parse_arx` move the pointer to the current document from parameter to class attribute, since most of the functions don't use it anyway (i.e. to silence compiler warnings)
- make most of the internal parsing functions in `parse_arx` private
- convert some loops to range-for

helper function and dependency changes
- use `lttoolbox/xml_walk_util.h` for iterating over a node's children and converting attribute values to `UString`

### CODE REVIEW

File: src/parse_arx.cc [hunk 1]
```diff
@@ -28,22 +28,20 @@
 #include <iostream>
 #include <lttoolbox/xml_parse_util.h>
 
-void print_tags(vector<wstring> input)
+void print_tags(vector<UString> input)
```
Comment (TinoDidriksen) [inline src/parse_arx.cc:31]: We should check everywhere vectors are passed by value. That's needlessly expensive.

File: src/pattern_arx.cc [hunk 2]
```diff
@@ -61,66 +63,70 @@ int contains_any(vector<wstring> tags, vector<wstring> candidates)
...
-   {
-      *p = towlower(*p);
-   }
+	UString temp;
+	size_t i = 0;
+	size_t len = s.size();
+	UChar32 c;
+	while (i < len) {
+		U16_NEXT(s, i, len, c);
+		temp += u_tolower(c);
```
Comment (TinoDidriksen) [inline src/pattern_arx.cc:74]: Nope, sorry, that's invalid. Correct Unicode case folding requires looking at surrounding code points, so this needs the full `u_strToLower()` https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/ustring_8h.html#a1b516af424a82c73421754e63306b4b6

**Comment by khannatanmai** [comment]
@mr-martian It feels like you're changing a lot more than just using ICU (at a glance, I could be wrong), and while improving the code is always great, changing too much in one PR makes it really hard to ensure no regression.

**Comment by mr-martian** [comment]
ICU-related changes
- replace all uses of `std::wstring` and associated types with `UString`
- use `lttoolbox/input_file.h` to read UTF-8 stream with nulls, since ICU can't do that directly
- use lttoolbox helper functions for manipulating string case

efficiency/readability changes
- pass containers by `const` reference whenever possible
- in `parse_arx` move the pointer to the current document from parameter to class attribute, since most of the functions don't use it anyway (i.e. to silence compiler warnings)
- make most of the internal parsing functions in `parse_arx` private
- convert some loops to range-for

changes departing from my original mandate
- use `lttoolbox/xml_walk_util.h` for iterating over a node's children and converting attribute values to `UString`

The first group touches everything by necessity and the second should have the effect of giving optimization hints to the compiler without changing structure.

The third group started out as updating string handling, but I had a helper for iterating over a node's children, and I decided that it made XML-handling code so much nicer that I was just going to use it everywhere.

**Pull request merged**